File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.*;

public class Main {

    static int N;
    static int cnt;
    static int[] board;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        board = new int[N];
        System.out.println(nQueen(0));
    }

    public static int nQueen(int row) {
        if (row == N) {
            cnt++;
            return cnt;
        }

        for (int i = 0; i < N; i++) {
            board[row] = i;

            if (isPromising(row)) {
                nQueen(row + 1);
            }
        }

        return cnt;
    }

    public static boolean isPromising(int row) {
        for (int i = 0; i < row; i++) {
            if (board[row] == board[i] || row - i == Math.abs(board[row] - board[i])) {
                return false;
            }
        }
        return true;
    }
}

File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <iostream>
#include <vector>

using namespace std;

int N = 0;
int cnt = 0;
vector<int> v;

int nQueen(int row);
bool check(int row);

int main() {
    cin >> N;
    v.resize(N);
    cout << nQueen(0);
}

int nQueen(int row) {
    if (row == N) {
        cnt++;
        return cnt;
    }
    
    for (int i = 0; i < N; i++) {        
        v[row] = i;
        if (check(row)) {
            nQueen(row+1);
        }
    }
    return cnt;
}

bool check(int row) {
    for (int i = 0; i < row; i++) {
        if (v[row] == v[i] || row - i == abs(v[row] - v[i])) {
            return false;
        }
    }
    return true;
}